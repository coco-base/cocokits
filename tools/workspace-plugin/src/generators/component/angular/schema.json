{
  "$schema": "https://json-schema.org/schema",
  "$id": "AngularComponent",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "project": {
      "type": "string",
      "description": "The project where the component will be located.",
      "x-prompt": "Which project will this component be located in?",
      "x-dropdown": "projects"
    },
    "subDirectory": {
      "type": "string",
      "description": "Sub-folder inside the 'lib' directory of target project where the component will be created.",
      "x-prompt": "What is the sub-folder 'lib' directory of target project that should contain the new component?",
      "default": "./"
    },
    "story": {
      "type": "boolean",
      "description": "Whether generate story of the component for storybook",
      "x-prompt": "Is the component has story for storybook?",
      "default": true
    }
  },
  "required": ["name", "project"]
}
