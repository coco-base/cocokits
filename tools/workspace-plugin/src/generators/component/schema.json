{
  "$schema": "https://json-schema.org/schema",
  "$id": "Component",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the component."
    },
    "project": {
      "type": "string",
      "description": "The project where the component will be located."
    },
    "story": {
      "type": "boolean",
      "description": "Whether generate story of the component for storybook",
      "default": true
    },
    "export": {
      "type": "boolean",
      "description": "Whether the component should be exported from the library",
      "default": true
    },
    "directory": {
      "type": "string",
      "description": "The directory for creating the generator file is determined by the `--directoryFormat` option. If set to `as-provided`, the path is relative to the current working directory. Otherwise, it is relative to the library root.",
      "default": "/"
    },
    "directoryFormat": {
      "type": "string",
      "description": "Directory configuration. When set to `as-provided`, the generator file is created in the given directory, relative to the current working directory, independent of the project structure. If set to `root`, the file is created relative to the library root, incorporating the project structure.",
      "enum": ["as-provided", "root"],
      "default": "as-provided"
    }
  },
  "required": ["name", "project"]
}
