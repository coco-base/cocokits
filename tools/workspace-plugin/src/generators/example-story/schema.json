{
  "$schema": "https://json-schema.org/schema",
  "$id": "ExampleStory",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "the name of example story. (e.g. back-button)"
    },
    "category": {
      "type": "string",
      "description": "the category of example story. It's the story Sidenav label. (e.g. button)"
    },
    "library": {
      "type": "string",
      "description": "the library of example story. (e.g. button, form-field)"
    },
    "uiComponentName": {
      "type": "string",
      "description": "the name of the UI component. (e.g. Button, IconButton)"
    },
    "angularInline": {
      "type": "boolean",
      "description": "use Angular inline template and styles",
      "default": true
    },
    "reactInline": {
      "type": "boolean",
      "description": "create a style file for React with styled-components",
      "default": true
    }
  },
  "required": ["name", "category", "library", "uiComponentName"]
}
