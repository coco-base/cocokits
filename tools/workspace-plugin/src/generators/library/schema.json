{
  "$schema": "https://json-schema.org/schema",
  "$id": "Library",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "framework": {
      "type": "string",
      "description": "The framework of library",
      "enum": ["angular", "react", "web", "html", "shared"],
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": "Select the framework of library"
    },
    "type": {
      "type": "string",
      "description": "The type of library",
      "enum": ["app", "ui", "theme", "util"],
      "$default": {
        "$source": "argv",
        "index": 2
      },
      "x-prompt": "Select the type of library"
    }
  },
  "required": ["name", "framework", "type"]
}
