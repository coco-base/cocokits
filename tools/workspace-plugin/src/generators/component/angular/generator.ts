import { formatFiles, generateFiles, names, Tree } from '@nx/devkit';
import path from 'path';

import { AngularComponentGeneratorSchema } from './schema';
import { ComponentFramework, ComponentGeneratorOptions } from '../component-generator.model';
import {
  getComponentImportPath,
  getComponentSelector,
  getProjectRoot,
  updateIndexFile,
} from '../component-generator.util';

export default async function angularComponentGenerator(tree: Tree, options: AngularComponentGeneratorSchema) {
  const componentGeneratorOptions: ComponentGeneratorOptions = {
    ...options,
    ...names(options.name),
    projectRoot: getProjectRoot(tree, options.project, ComponentFramework.Angular),
    framework: ComponentFramework.Angular,
  };

  generateComponent(tree, componentGeneratorOptions);

  if (componentGeneratorOptions.story) {
    generateStoryComponent(tree, componentGeneratorOptions);
  }

  updateIndexFile(tree, componentGeneratorOptions);

  await formatFiles(tree);
}

function generateComponent(tree: Tree, componentOptions: ComponentGeneratorOptions) {
  generateFiles(
    tree,
    path.join(__dirname, 'templates', 'component'),
    path.join(componentOptions.projectRoot, 'src', 'lib', componentOptions.subDirectory, componentOptions.fileName),
    {
      ...componentOptions,
      selector: getComponentSelector(componentOptions),
    }
  );
}

function generateStoryComponent(tree: Tree, componentOptions: ComponentGeneratorOptions) {
  generateFiles(
    tree,
    path.join(__dirname, 'templates', 'storybook'),
    path.join(componentOptions.projectRoot, 'stories', componentOptions.fileName),
    {
      ...componentOptions,
      componentImportPath: getComponentImportPath(componentOptions),
    }
  );
}
