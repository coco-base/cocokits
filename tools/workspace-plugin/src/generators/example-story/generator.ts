import { formatFiles, generateFiles, names, Tree } from '@nx/devkit';
import * as path from 'path';

import { toTitleCase } from '@cocokits/common-utils';

import { ExampleStoryGeneratorOptions } from './options';
import { ExampleStoryGeneratorSchema } from './schema';

export async function exampleStoryGenerator(tree: Tree, schema: ExampleStoryGeneratorSchema) {
  const paredNames = names(schema.name);
  const options: ExampleStoryGeneratorOptions = {
    ...paredNames,
    titleName: toTitleCase(paredNames.name),
    rawName: schema.name,
    library: schema.library,
    uiComponentName: schema.uiComponentName,
    angularInline: schema.angularInline,
    reactInline: schema.reactInline,
  };

  const reactExampleFolderPath = `apps/react-kits-doc/stories/ui-components/${options.library}/${options.uiComponentName}/examples`; // back-button/component.tsx
  const angularExampleFolderPath = `apps/angular-kits-doc/stories/ui-components/${options.library}/${options.uiComponentName}/examples`; // back-button/component.ts
  const configExampleFolderPath = `packages/internal/common-kits-doc/src/examples-config/${options.uiComponentName}`; // back-button.config.ts

  generateFiles(
    tree,
    path.join(__dirname, `files/${options.reactInline ? 'react-inline' : 'react'}`),
    reactExampleFolderPath,
    options
  );
  addImport(tree, reactExampleFolderPath, false, options);

  generateFiles(
    tree,
    path.join(__dirname, `files/${options.angularInline ? 'angular-inline' : 'angular'}`),
    angularExampleFolderPath,
    options
  );
  addImport(tree, angularExampleFolderPath, true, options);

  generateFiles(tree, path.join(__dirname, `files/config`), configExampleFolderPath, options);

  await formatFiles(tree);
}

function findIndexStory(tree: Tree, startDir: string): string | null {
  let currentDir = startDir;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const candidateTs = path.join(currentDir, 'index.stories.ts');
    const candidateTsx = path.join(currentDir, 'index.stories.tsx');

    if (tree.exists(candidateTs)) {
      return candidateTs;
    }

    if (tree.exists(candidateTsx)) {
      return candidateTsx;
    }

    const parent = path.dirname(currentDir);
    if (parent === currentDir) {
      return null;
    }
    currentDir = parent;
  }
}

function addImport(tree: Tree, exampleFolderPath: string, isAngular: boolean, options: ExampleStoryGeneratorOptions) {
  const indexStoryPath = findIndexStory(tree, exampleFolderPath);

  if (!indexStoryPath || !tree.exists(indexStoryPath)) {
    return;
  }

  let content = tree.read(indexStoryPath, 'utf-8');

  const lastMatch = findLastExportLine(content);

  if (lastMatch) {
    const newLine = `export { ${options.className}${isAngular ? '' : 'Story'} } from './examples/${options.fileName}/index.example.stories';`;
    content = content.replace(lastMatch, `${lastMatch}\n${newLine}`);
  } else {
    console.warn(
      `Could not find a place to export the generate story in ${indexStoryPath}. You will need to add the export manually.`
    );
  }

  tree.write(indexStoryPath, content);
}

function findLastExportLine(content: string): string | null {
  let lastMatch: string | null = null;
  let match: RegExpExecArray | null;

  // Matches any line containing from './examples/ plus any characters after
  // and captures the entire line
  const regexExample = /^(.*from\s+['"]\.\/examples\/.*['"].*)$/gm;

  // Find the last occurrence of the matching line
  while ((match = regexExample.exec(content)) !== null) {
    lastMatch = match[1];
  }

  if (lastMatch) {
    return lastMatch;
  }

  // Matches any line containing from './examples/ plus any characters after
  // and captures the entire line
  const regexOverview = /^(.*from\s+['"]\.\/overview\/.*['"].*)$/gm;

  // Find the last occurrence of the matching line
  while ((match = regexOverview.exec(content)) !== null) {
    lastMatch = match[1];
  }

  return lastMatch;
}

export default exampleStoryGenerator;
