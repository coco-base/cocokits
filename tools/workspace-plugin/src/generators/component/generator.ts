import { formatFiles, Tree } from '@nx/devkit';

import { ComponentGeneratorSchema } from './schema';
import { angularComponentGenerator } from './src/angular-component-generator';
import { reactComponentGenerator } from './src/react-component-generator';
import { getOptions } from './utils/get-options';
import { LibraryFramework } from '../generator.model';

export async function componentGenerator(tree: Tree, schema: ComponentGeneratorSchema) {
  const options = getOptions(tree, schema);

  switch (options.framework) {
    case LibraryFramework.Angular:
      await angularComponentGenerator(tree, options);
      break;

    case LibraryFramework.React:
      await reactComponentGenerator(tree, options);
      break;

    default:
      throw new Error(`The generator doesn't support create component for '${options.framework}' framework right now`);
  }

  await formatFiles(tree);
}

export default componentGenerator;
