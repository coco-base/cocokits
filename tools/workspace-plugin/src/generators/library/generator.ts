import { generateFiles, names, Tree } from '@nx/devkit';
import { LibraryGeneratorOptions, LibraryGeneratorSchema } from './schema';
import { CckLibraryFramework, CckLibraryType } from './library-generator.model';
import type { Schema as NgSchema } from '@nx/angular/src/generators/library/schema';
import { Linter } from '@nx/eslint';
import { libraryGenerator as ngLibraryGenerator, UnitTestRunner } from '@nx/angular/generators';
import * as path from 'path';

export default async function libraryGenerator(tree: Tree, options: LibraryGeneratorSchema) {
  const libraryOptions: LibraryGeneratorOptions = {
    ...options,
    ...names(options.name),
  };

  switch (options.framework) {
    case CckLibraryFramework.Angular:
      await angularLibraryGenerator(tree, libraryOptions);
      break;

    case CckLibraryFramework.React:
      await reactLibraryGenerator(tree, libraryOptions);
      break;

    case CckLibraryFramework.Web:
      await webLibraryGenerator(tree, libraryOptions);
      break;

    case CckLibraryFramework.Html:
      await htmlLibraryGenerator(tree, libraryOptions);
      break;

    case CckLibraryFramework.Shared:
      await angularLibraryGenerator(tree, libraryOptions);
      await reactLibraryGenerator(tree, libraryOptions);
      await webLibraryGenerator(tree, libraryOptions);
      await htmlLibraryGenerator(tree, libraryOptions);
      break;
  }
}

async function angularLibraryGenerator(tree: Tree, options: LibraryGeneratorOptions) {
  const ngSchema: NgSchema = {
    name: options.name,
    directory: getNxLibraryDirectory(options),
    buildable: true,
    publishable: true,
    importPath: getNxImportPath(options),
    tags: getNxTags(options),
    linter: Linter.None,
    unitTestRunner: UnitTestRunner.None,
    skipModule: true,
    skipTests: true,
    standalone: false,
  };

  const libPath = `${ngSchema.directory}/${ngSchema.name}`;
  const eslintPath = `tools/eslint`;
  const relativePathToEslint = path.relative(libPath, eslintPath);

  await ngLibraryGenerator(tree, ngSchema);
  generateFiles(tree, path.join(__dirname, 'templates', 'angular'), path.join(libPath), {
    baseEslintDir: relativePathToEslint,
  });
}

async function reactLibraryGenerator(tree: Tree, options: LibraryGeneratorOptions) {
  console.log('library generator for react is not implemented.');
}

async function webLibraryGenerator(tree: Tree, options: LibraryGeneratorOptions) {
  console.log('library generator for web is not implemented.');
}

async function htmlLibraryGenerator(tree: Tree, options: LibraryGeneratorOptions) {
  console.log('library generator for html is not implemented.');
}

function getNxTags(options: LibraryGeneratorOptions) {
  return `type:${options.type}, framework:${options.framework}`;
}

function getNxImportPath(options: LibraryGeneratorOptions) {
  switch (options.type) {
    case CckLibraryType.Theme:
      return `@coco-kits/theme-${options.fileName}`;
    case CckLibraryType.App:
      return `@coco-kits/app-${options.fileName}`;
    case CckLibraryType.Ui:
      return `@coco-kits/${options.framework}-${options.fileName}`;
    case CckLibraryType.Util:
      return `@coco-kits/${options.framework}-utils-${options.fileName}`;
  }
}

function getNxLibraryDirectory(options: LibraryGeneratorOptions) {
  switch (options.type) {
    case CckLibraryType.Theme:
      return `packages/themes`;
    case CckLibraryType.App:
      return `apps`;
    case CckLibraryType.Ui:
      return `packages/${options.framework}`;
    case CckLibraryType.Util:
      return `packages/${options.framework}/utils`;
  }
}
