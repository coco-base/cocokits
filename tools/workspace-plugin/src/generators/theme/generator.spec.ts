import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree, readProjectConfiguration } from '@nx/devkit';

import { themeGenerator } from './generator';
import { ThemeGeneratorSchema } from './schema';
import { beforeEach, describe, expect, it } from 'vitest';

describe('theme generator', () => {
  let tree: Tree;
  const options: ThemeGeneratorSchema = { name: 'test' };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await themeGenerator(tree, options);
    const config = readProjectConfiguration(tree, 'test');
    expect(config).toBeDefined();
    expect(tree.read(`libs/test/src/index.ts`, 'utf-8')).toMatchSnapshot();
  });
});
