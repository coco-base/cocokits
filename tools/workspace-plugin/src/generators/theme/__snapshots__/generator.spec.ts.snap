// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`theme generator > should run successfully 1`] = `
"const variable = 'test';
"
`;
